source/extensions/filters/network/ratelimit/ratelimit.h

```diff
@@ -51,15 +54,21 @@ class Config {
   Runtime::Loader& runtime() { return runtime_; }
   const InstanceStats& stats() { return stats_; }
   bool failureModeAllow() const { return !failure_mode_deny_; };
+  Formatter::FormatterImpl substitutionFormatter() { return Formatter::FormatterImpl(""); }
+  void applySubstitutionFormatter(std::vector<RateLimit::Descriptor> original_descriptors,
+                                  StreamInfo::StreamInfo& stream_info);
 
 private:
   static InstanceStats generateStats(const std::string& name, Stats::Scope& scope);
-
   std::string domain_;
   std::vector<RateLimit::Descriptor> descriptors_;
+  std::vector<std::shared_ptr<Formatter::FormatterImpl>> substitution_formatters_;
```

Comment:
nit: I think you can use unique_ptr here.

Reply:
Other similar use cases in this project are using shared_ptr instead of unique_ptr and I"m not the best C++ developer as you can tell. I will keep as is.

Reply:
I'm pretty sure you can use unique_ptr here, so let's give that a try.